document-generator-accounts/src/main/java/uk/gov/companieshouse/document/generator/accounts/handler/accounts/AccountsHandlerImpl.java

```diff
@@ -1,13 +1,55 @@
 package uk.gov.companieshouse.document.generator.accounts.handler.accounts;
 
+import static uk.gov.companieshouse.document.generator.accounts.AccountsDocumentInfoServiceImpl.MODULE_NAME_SPACE;
+
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
+import uk.gov.companieshouse.api.model.accounts.Accounts;
+import uk.gov.companieshouse.document.generator.accounts.AccountType;
+import uk.gov.companieshouse.document.generator.accounts.LinkType;
+import uk.gov.companieshouse.document.generator.accounts.exception.HandlerException;
+import uk.gov.companieshouse.document.generator.accounts.exception.ServiceException;
+import uk.gov.companieshouse.document.generator.accounts.service.AccountsService;
 import uk.gov.companieshouse.document.generator.interfaces.model.DocumentInfoResponse;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
 
 @Component
 public class AccountsHandlerImpl implements AccountsHandler  {
 
+    private static final Logger LOG = LoggerFactory.getLogger(MODULE_NAME_SPACE);
+
+    @Autowired
+    AccountsService accountsService;
+
     @Override
-    public DocumentInfoResponse getAccountsData(String resourceLink) {
+    public DocumentInfoResponse getAccountsData(String resourceLink) throws HandlerException {
+        Accounts accounts;
+
+        try {
+            accounts = accountsService.getAccounts(resourceLink);
+        } catch (ServiceException e) {
+            throw new HandlerException(e.getMessage(), e.getCause());
+        }
+
+        // TODO: The accountsType variable shall be used in the implementation of abridged logic
+        AccountType accountsType = getAccountType(accounts);
+
         return new DocumentInfoResponse();
     }
+
+
+    /**
+     * Get the account type from the account link within links
+     *
+     * @return accountsData - accounts data
+     */
+    private AccountType getAccountType(Accounts accountsData) throws HandlerException {
+        return accountsData.getLinks().keySet()
```

Comment:
you shouldnt iterate over the keyset to check if the key exists in a map. Just use Optional instead, if the key get returns a null then it will still throw the exception you expect.
```
return Optional.ofNullable(accountsData.getLinks())
                          .map(links -> links.get(LinkType.SELF.getLink()))
                          .map(AccountType::getAccountType)
                          .orElseThrow(...);
```

Reply:
Optional doesn't work as we are searching for LinkTypes that aren't SELF link.

Reply:
yep my bad.